### PR TITLE
Emit fixes local to the result, not as an aggregate

### PR DESCRIPTION
**Pull request by kristof-mattei**
Fix for https://github.com/returntocorp/semgrep/issues/5421

Testing https://github.com/returntocorp/semgrep/blob/5dbf6dd47ab4fda6e9dc8dec42b90fce24d1fd82/semgrep/tests/e2e/snapshots/test_output/test_sarif_output_with_autofix/results.sarif against https://github.com/oasis-tcs/sarif-spec/blob/master/Documents/CommitteeSpecifications/2.1.0/sarif-schema-2.1.0.json validates 100%.

I don't want to use that url though as the $schema in the files we generate, as that is not the spec. 

PR checklist:

- [x] Documentation is up-to-date
- [x] Changelog is up-to-date
   didn't update as the prefix 'fixes' is still under 'unreleased': https://github.com/returntocorp/semgrep/blob/develop/CHANGELOG.md#added
- [x] Change has no security implications (otherwise, ping security team)

### CODE REVIEW

File: semgrep/semgrep/formatter/sarif.py [hunk 1]
```diff
@@ -67,7 +72,7 @@ def _rule_match_to_sarif_fix(rule_match: RuleMatch) -> Optional[Mapping[str, Any
                                 "endLine": rule_match.end.line,
                                 "endColumn": rule_match.end.col,
                             },
-                            "insertedContent": {"text": fixed_lines},
+                            "insertedContent": fixed_lines,
```
Comment (ajbt200128) [inline semgrep/semgrep/formatter/sarif.py]: Why remove text? See some examples [here](https://docs.oasis-open.org/sarif/sarif/v2.0/csprd01/sarif-v2.0-csprd01.html#_Toc517436085 ), it seems like it makes sense to keep it

File: semgrep/semgrep/formatter/sarif.py [hunk 2]
```diff
@@ -185,15 +190,6 @@ def format(
...
-        rule_matches, rule_matches_fixes = tee(rule_matches)
-        fixes = [
-            self._rule_match_to_sarif_fix(rule_match)
-            for rule_match in rule_matches_fixes
-        ]
-        # Filter out rule matches w/no fixes
-        fixes = list(filter(None, fixes))
-
         output_dict = {
             "$schema": "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json",
```
Comment (ajbt200128) [inline semgrep/semgrep/formatter/sarif.py]: Can we update this broken link to https://docs.oasis-open.org/sarif/sarif/v2.0/csprd01/schemas/sarif-schema.json ?

File: semgrep/tox.ini [hunk 3]
```diff
@@ -6,7 +6,7 @@ envlist = py37, py38, py39
 allowlist_externals = pipenv
 commands =
     pipenv install --sequential --dev
-    pipenv run pytest --ignore=tests/qa/test_public_repos.py -n auto -vv --tb=short
+    pipenv run pytest --ignore=tests/qa/test_public_repos.py -n auto -vv --tb=short {posargs}
```
Comment (ajbt200128) [inline semgrep/tox.ini]: Please remove this line, we do want to run these QA tests, and this isn't relevant to the scope of the issue

**Review by ajbt200128 — CHANGES_REQUESTED** [review]

File: semgrep/semgrep/formatter/sarif.py (hunk 1 above)
Comment (kristof-mattei) [reply semgrep/semgrep/formatter/sarif.py]: Schema validation fails if we do.

File: semgrep/semgrep/formatter/sarif.py (hunk 2 above)
Comment (kristof-mattei) [reply semgrep/semgrep/formatter/sarif.py]: Will do.

File: semgrep/tox.ini (hunk 3 above)
Comment (kristof-mattei) [reply semgrep/tox.ini]: Will do.

File: semgrep/semgrep/formatter/sarif.py (hunk 1 above)
Comment (kristof-mattei) [reply semgrep/semgrep/formatter/sarif.py]: Update: https://docs.oasis-open.org/sarif/sarif/v2.0/csprd01/schemas/sarif-schema.json it is in this one.

But it is not in this one: https://github.com/oasis-tcs/sarif-spec/blob/master/Documents/CommitteeSpecifications/2.1.0/sarif-schema-2.1.0.json

Comment (kristof-mattei) [reply semgrep/semgrep/formatter/sarif.py]: Let me see if we can check which validation GitHub uses.

**Comment by kristof-mattei** [comment]
@ajbt200128 trying with https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/schemas/sarif-schema-2.1.0.json, which is the latest version.

**Comment by ajbt200128** [comment]
> @ajbt200128 trying with https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/schemas/sarif-schema-2.1.0.json, which is the latest version.

Great, if it works, make sure to regenerate the tests so the output contains the new schema url :)

**Comment by kristof-mattei** [comment]
@ajbt200128 had to join `text` into a single string as the scheme doesn't allow for `[]`. Ran tests, merged in changes. LMK.

**Comment by ajbt200128** [comment]
> @ajbt200128 had to join `text` into a single string as the scheme doesn't allow for `[]`. Ran tests, merged in changes. LMK.

Good catch, can you escape the newlines though instead of removing them? That'd be in accordant w/ the json specs. After that we should be good

**Comment by kristof-mattei** [comment]
@ajbt200128 I didn't remove them. `text` is an array when it comes from `semgrep`, and then they are joined with `\n`:
https://github.com/returntocorp/semgrep/pull/5425/files#diff-9a6ebaf633a3e083e4a7792fb44df39b96d443dc8641055c8cabc5f50bf8f3baR75

**Comment by ajbt200128** [comment]
@kristof-mattei my bad, I misread, but either way, they should be joined with escaped new lines, as per the json spec

**Comment by kristof-mattei** [comment]
@ajbt200128 they are escaped in the rendered output:

```
❯ python3
Python 3.10.4 (main, Mar 23 2022, 20:25:24) [GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import json
>>> foo = ["hello", "world"]
>>> bar = { "foo": "\n".join(foo) }
>>> bar
{'foo': 'hello\nworld'}
>>> json.dumps(bar)
'{"foo": "hello\\nworld"}'
>>>
```

**Comment by ajbt200128** [comment]
great!

**Review by ajbt200128 — APPROVED** [review]
Looks good!

**Comment by aryx** [comment]
@ajbt200128 can you push this on the finish line?

**Comment by michael-schwarz** [comment]
It would be nice if this could be merged soon, it's quite annoying to have the CI jobs fail on every commit because of this. :smile:

**Pull request merged**